src/util/subscribeToResult.ts

```diff
@@ -68,15 +67,18 @@ export function subscribeToResult<T>(outerSubscriber: OuterSubscriber<any, any>,
         break;
       }
     } while (true);
-  } else if (typeof result[$$observable] === 'function') {
+  } else if (result && typeof result[$$observable] === 'function') {
     const obs = result[$$observable]();
     if (typeof obs.subscribe !== 'function') {
-      destination.error(new Error('invalid observable'));
+      destination.error(new TypeError('Provided object does not correctly implement Symbol.observable'));
     } else {
       return obs.subscribe(new InnerSubscriber(outerSubscriber, outerValue, outerIndex));
     }
   } else {
-    destination.error(new TypeError('unknown type returned'));
+    const value = isObject(result) ? 'an invalid object' : `'${result}'`;
+    const msg = `You provided ${value} where a stream was expected.`
+      + ' You can provide an Observable, Promise, Array, or Iterable.';
```

Comment:
It might be nitpicking, but one could return an Rx.Observable or a "lowercase-o observable" which is any object that implements `Symbol.observable`. How do we differentiate those two things there? Maybe `Observable, Promise, Array, Iterable or object implementing Symbol.observable`?